src/main/java/org/opensearch/agent/tools/CreateAnomalyDetectorToolEnhanced.java

```diff
@@ -0,0 +1,1402 @@
+                                    )
+                                );
+                        }
+                        return;
+                    }
+                    suggestHyperParametersPhase(detector, tenantId, maxRetries, listener);
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    log.error("Detector validation API failed: {}", e.getMessage(), e);
+                    if (currentRetry < MAX_DETECTOR_VALIDATION_RETRIES) {
+                        retryDetectorValidation(
+                            suggestions,
+                            mappingContext,
+                            e.getMessage(),
+                            tenantId,
+                            maxRetries,
+                            currentRetry + 1,
+                            listener
+                        );
+                    } else {
+                        listener
+                            .onFailure(
+                                new RuntimeException(
+                                    "Detector validation failed after " + MAX_DETECTOR_VALIDATION_RETRIES + " retries: " + e.getMessage()
+                                )
+                            );
+                    }
+                }
+            });
+
+        } catch (Exception e) {
+            log.error("Error building detector: {}", e.getMessage(), e);
+            if (currentRetry < MAX_FORMAT_FIX_RETRIES) {
+                retryDetectorValidation(suggestions, mappingContext, e.getMessage(), tenantId, maxRetries, currentRetry + 1, listener);
+            } else {
+                listener.onFailure(e);
+            }
+        }
+    }
+
+    private <T> void retryDetectorValidation(
+        Map<String, String> originalSuggestions,
+        MappingContext mappingContext,
+        String validationError,
+        String tenantId,
+        int maxRetries,
+        int currentRetry,
+        ActionListener<T> listener
+    ) {
+        String indexName = originalSuggestions.get(OUTPUT_KEY_INDEX);
+
+        StringBuilder contextBuilder = new StringBuilder();
+        contextBuilder.append("You are creating an anomaly detector for the following index:\n\n");
+        contextBuilder.append("Index: ").append(indexName).append("\n");
+        contextBuilder.append("Available fields:\n");
+        for (Map.Entry<String, String> field : mappingContext.filteredMapping.entrySet()) {
+            contextBuilder.append("- ").append(field.getKey()).append(": ").append(field.getValue()).append("\n");
+        }
+        contextBuilder.append("Available date fields: ").append(String.join(", ", mappingContext.dateFields)).append("\n\n");
+
+        String fixPrompt = createFixPrompt(originalSuggestions, validationError);
+        String fullPrompt = contextBuilder.toString() + fixPrompt;
+
+        callLLM(fullPrompt, tenantId, ActionListener.wrap(fixedResponse -> {
```

Comment:
Lots of duplicated code between `retryDetectorValidation()` and `retryModelValidation`, consider a helper method for the common parts.